src/bsoncxx/types/value.cpp

```diff
@@ -40,15 +45,28 @@ value::value(const value& rhs) {
 }
 
 value& value::operator=(const value& rhs) {
-    _type = rhs._type;
+    if(_type && !rhs._type) {
```

Comment:
From [cppreference](http://en.cppreference.com/w/cpp/language/union): "If members of a union are classes with user-defined constructors and destructors, to switch the active member, explicit destructor and placement new are generally needed"

I don't think we've defined any custom dtors for the b_* types now, but we call ~b_*type() in ~value().  It's probably not necessary to call the dtors because we have default destructors but this could change in the future, and it might be better to just add this now.